### PR TITLE
[export] fix unbacked range deserialization

### LINKED ISSUE #151809: [export] deserialization for unbacked ranges is wrong
### 🐛 Describe the bug

ShapeEnv range info is wrong for unbacked symbols after we deserialize, with lower bound of 2:
```
import io
import torch
from torch.export import export, save, load

class Foo(torch.nn.Module):
    def forward(self, x):
        n = x.item()
        return torch.empty(n)

ep = export(Foo(), (torch.tensor([5]),))
buffer = io.BytesIO()
save(ep, buffer)
buffer.seek(0)
loaded_ep = load(buffer)

# pre-serialize ep
print("pre-serialize")
shape_env = torch._guards.detect_fake_mode([
    node.meta.get("val") for node in ep.graph.nodes
]).shape_env
print(shape_env.var_to_range)

# deserialized ep
print("deserialized")
shape_env = torch._guards.detect_fake_mode([
    node.meta.get("val") for node in loaded_ep.graph.nodes
]).shape_env
print(shape_env.var_to_range)
```

we get:
```
pre-serialize
{u0: VR[0, int_oo]}
deserialized
{u0: VR[2, int_oo]}
```

This happens because we were blindly clamping lower bounds for all symbols (this was intended just for backed symbols, so users could specify min=0 or 1): https://github.com/pytorch/pytorch/blob/0f8613bf5cbdd7a2af5c46e6fa1adda35c69db8d/torch/_export/serde/serialize.py#L2171

But this was conveniently helping us get past 0/1 data-dependent errors when deserializing tensor values (in empty_strided calls), which were never exposed. A more correct fix could be to save and load size-like info, and deserialize node-by-node, storing runtime asserts in the ShapeEnv as needed. Or we could just start serializing the ShapeEnv in full, so we stop running into such information loss issues.

### Versions

Collecting environment information...
PyTorch version: 2.8.0a0+git1a48382
Is debug build: False
CUDA used to build PyTorch: 12.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-5)
Clang version: Could not collect
CMake version: version 3.30.2
Libc version: glibc-2.34

Python version: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.4.3-0_fbk14_hardened_2601_gcd42476b84e9-x86_64-with-glibc2.34
Is CUDA available: False
CUDA runtime version: 12.0.140
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: 
GPU 0: NVIDIA H100
GPU 1: NVIDIA H100

Nvidia driver version: 550.90.07
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      52 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             92
On-line CPU(s) list:                0-91
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 9654 96-Core Processor
CPU family:                         25
Model:                              17
Thread(s) per core:                 1
Core(s) per socket:                 92
Socket(s):                          1
Stepping:                           1
BogoMIPS:                           4792.79
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core invpcid_single ssbd ibrs ibpb stibp vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean pausefilter pfthreshold v_vmsave_vmload vgif avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid fsrm arch_capabilities
Virtualization:                     AMD-V
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          5.8 MiB (92 instances)
L1i cache:                          5.8 MiB (92 instances)
L2 cache:                           46 MiB (92 instances)
L3 cache:                           1.4 GiB (92 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-91
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:           Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] adam-atan2-pytorch==0.1.1
[pip3] alphafold3-pytorch==0.6.6
[pip3] bert_pytorch==0.0.1a4
[pip3] ema-pytorch==0.7.3
[pip3] executorch==0.4.0.dev20240809+cpu
[pip3] flake8==7.1.1
[pip3] frame-averaging-pytorch==0.1.2
[pip3] lion-pytorch==0.2.2
[pip3] mypy==1.9.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.1.3.1
[pip3] nvidia-cuda-cupti-cu12==12.1.105
[pip3] nvidia-cuda-nvrtc-cu12==12.1.105
[pip3] nvidia-cuda-runtime-cu12==12.1.105
[pip3] nvidia-cudnn-cu12==8.9.2.26
[pip3] nvidia-cufft-cu12==11.0.2.54
[pip3] nvidia-curand-cu12==10.3.2.106
[pip3] nvidia-cusolver-cu12==11.4.5.107
[pip3] nvidia-cusparse-cu12==12.1.0.106
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.18.1
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.1.105
[pip3] onnx==1.16.1
[pip3] onnxruntime==1.18.0
[pip3] onnxscript==0.3.0.dev20250225
[pip3] open-clip-torch==2.24.0
[pip3] optree==0.13.1
[pip3] pytorch-labs-segment-anything-fast==0.2
[pip3] pytorch-lightning==2.0.7
[pip3] pytorch_sphinx_theme==0.0.24
[pip3] pytorch-triton==3.0.0+45fff310c8
[pip3] rotary-embedding-torch==0.8.5
[pip3] torch==2.8.0a0+git1a48382
[pip3] torch_geometric==2.4.0
[pip3] torch-mlir==20241017.255
[pip3] torch-stoi==0.2.1
[pip3] torch_tensorrt==2.6.0
[pip3] torchao==0.10.0+git7d879462
[pip3] torchaudio==2.6.0.dev20250131+cpu
[pip3] torchdiffeq==0.2.4
[pip3] torchmetrics==1.0.3
[pip3] torchrec==0.9.0a0+5e30669
[pip3] torchsde==0.2.6
[pip3] torchsr==1.0.4
[pip3] torchtext==0.18.0
[pip3] torchtune==0.5.0
[pip3] torchtyping==0.1.5
[pip3] torchvision==0.22.0a0+fab1188
[pip3] torchx==0.7.0
[pip3] triton==3.1.0
[conda] Could not collect

cc @chauhang @penguinwu @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

### PR DESCRIPTION
**Pull request by pianpwk**
Fixes https://github.com/pytorch/pytorch/issues/151809, by reading shape assertion nodes into ShapeEnv, and deferring instantiation of node example values, to be done node-by-node.

Differential Revision: D78588406

### CODE REVIEW

**Review by ydwu4 — APPROVED** [review]

File: torch/_export/serde/serialize.py [hunk 1]
```diff
@@ -1876,33 +1899,23 @@ def deserialize_graph(self, serialized_graph: Graph) -> torch.fx.Graph:
         # Handle the tensor metas.
         for name, tensor_value in serialized_graph.tensor_values.items():
             log.debug("[deserialize_tensor_meta] %s (input): %s", name, tensor_value)
-            meta_val = self.deserialize_tensor_meta(tensor_value)
-            log.debug("[deserialize_tensor_meta] %s (output): %s", name, meta_val)
-            self.serialized_name_to_meta[name] = meta_val
+            self.serialized_name_to_meta[name] = lambda v=tensor_value: self.deserialize_tensor_meta(v)
```
Comment (avikchaudhuri) [inline torch/_export/serde/serialize.py]: Have you checked Python semantics re: capture of `tensor_value` here? Hope it doesn't pick up values assigned later and instead does what you think it does...which is capture the value in *this* iteration?

**Review by avikchaudhuri — APPROVED** [review]

File: torch/_export/serde/serialize.py (hunk 1 above)
Comment (pianpwk) [reply torch/_export/serde/serialize.py]: Yeah, the default value here is for using the value at definition; doing `lambda: self.deserialize_tensor_meta(tensor_value)` was using the value at evaluation.

Comment (avikchaudhuri) [reply torch/_export/serde/serialize.py]: OK. Minor nit: could have used `functools.partial` as well I think for this purpose, the default value trick seems a bit obscure.
